tools/kprint/addBrackets.cpp

```diff
@@ -0,0 +1,359 @@
+ *
+ * s1: first sort
+ *
+ * s2: second sort
+ *
+ * returns: s1 <= s2
+ */
+bool lessThanEq(PrettyPrintData const& data, KORESort *s1, KORESort *s2) {
+  return *s1 == *s2 || (data.subsorts.count(s1) && data.subsorts.at(s1).count(s2));
+}
+
+/**
+ * Computes whether `inner` can appear at position `position` inside `outer`
+ * according to the rules of subsorting, priority, and associativity. The
+ * pattern does not actually have to exist at that position; we ask only
+ * whether it `can` appear there.
+ *
+ * outer: the outer pattern
+ *
+ * inner: the inner pattern
+ *
+ * position: a position within `outer` you wish to test
+ *
+ * data: pretty printing metadata
+ *
+ * returns: true if priority forbids `inner` to appear inside `outer` at that
+ * position.
+ */
+bool isPriorityWrong(KORECompositePattern *outer, KORECompositePattern *inner, int position, PrettyPrintData const& data) {
+  std::string outerName = outer->getConstructor()->getName();
+  std::string innerName = inner->getConstructor()->getName();
+  KORESort *innerSort = inner->getSort().get();
+  KORESort *outerSort = outer->getConstructor()->getArguments()[position].get();
+  if (!lessThanEq(data, innerSort, outerSort)) {
+    return true;
+  }
+  if (data.priorities.count(outerName) && data.priorities.at(outerName).count(innerName)) {
+    return true;
+  }
+  std::string terminals = data.terminals.at(outerName);
+  int terminalPos = getNTPositionInProd(terminals, position);
+  if (data.leftAssoc.count(outerName) && data.leftAssoc.at(outerName).count(innerName) && terminalPos == terminals.size() - 1) {
+    return true;
+  }
+  if (data.rightAssoc.count(outerName) && data.rightAssoc.at(outerName).count(innerName) && terminalPos == 0) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Computes whether a bracket needs to appear around `inner`.
+ *
+ * outer: the parent of `inner`
+ *
+ * leftCapture: the left capture of `inner`
+ *
+ * rightCapture: the right capture of `inner`
+ *
+ * inner: the pattern to test
+ *
+ * position: the position of `inner` within the nonterminals of `outer`
+ *
+ * data: pretty printing metadata
+ *
+ * returns: true if a bracket is required according to this algorithm. Roughly
+ * speaking, parentheses are not needed around associative symbols, constants,
+ * or patterns that appear between two terminals in their parent. Otherwise,
+ * parentheses are needed if priority or associativity forbids the term inside
+ * its parent, or if the term begins or ends with a nonterminal and priority
+ * and associativity do not forbid the term from appearing inside its left
+ * or right capture, respectively, and its left or right capture does not
+ * end or begin, respectively, with a terminal.
+ *
+ * Warning: this algorithm is *unsound*. There is no guarantee that after 
+ * it completes, the resulting term can be unparsed into a string that will
+ * parse unambiguously as the original AST. Indeed, the problem of inserting
```

Comment:
Do you have an example for unsoundness? It's likely not optimal, but I couldn't find a case where (say) it leaves out a parenthesis, or it produces ambiguity.

Reply:
I mean, I believe the obvious example is this:

```
syntax Exp ::= Exp "++" | "++" Exp | Exp Exp | "(" Exp ")" [bracket]

(1 ++) 2
```

This algorithm is not smart enough to insert a parentheses here.

Reply:
Thanks! Could you add it to the comment?